### PR TITLE
Add following APIs to facilitate DataSet conversion

### PR DESCRIPTION
**Pull request by cydrain**
* ConvertFromDataTypeIfNeeded()
 * ConvertToDataTypeIfNeeded()

### CODE REVIEW

**Review by Presburger — APPROVED** [review]

File: include/knowhere/utils.h [hunk 1]
```diff
@@ -127,6 +127,39 @@ data_type_conversion(const DataSet& src) {
...
+        return data_type_conversion<DataType, typename MockData<DataType>::type>(*ds);
+    }
+}
+
+// Convert DataSet from DataType to float
+template <typename DataType>
+inline std::shared_ptr<const DataSet>
+ConvertFromDataTypeIfNeeded(const DataSet& ds) {
+    if constexpr (std::is_same_v<DataType, typename MockData<DataType>::type>) {
+        return ds.shared_from_this();
```
Comment (zhengbuqian) [inline include/knowhere/utils.h:146]: this is very strange: we don't know who owns the dataset, how can we just create a shared ptr from it? This implicitly requires the passed in `ds` to be already managed by a shared ptr. This requirement is satisfied for now, but could lead to a hard to find bug in the future.

but this issue is not introduced by this PR, we can figure out a solution and put up a fix later.

Comment (cydrain) [reply include/knowhere/utils.h:146]: plan to file another PR to change all APIs' input parameter to shared_ptr to avoid this issue completely

**Pull request merged**